src/passes/OptimizeAddedConstants.cpp

```diff
@@ -123,12 +126,29 @@ template<typename P, typename T> class MemoryAccessOptimizer {
       // code may know that is valid, even if we can't. Only handle the
       // obviously valid case where an overflow can't occur.
       auto* c = curr->ptr->template cast<Const>();
-      uint32_t base = c->value.geti32();
-      uint32_t offset = curr->offset;
-      if (uint64_t(base) + uint64_t(offset) < (uint64_t(1) << 32)) {
-        c->value = c->value.add(Literal(uint32_t(curr->offset)));
-        curr->offset = 0;
-      }
+
+      auto optIfSatisfied = [&](Const* c) {
+        if (memory64) {
+          uint64_t base = c->value.geti64();
+          uint64_t offset = curr->offset;
+
+          uint64_t max = std::numeric_limits<uint64_t>::max();
+          bool overflow = (base > max - offset) || (offset > max - base);
```

Comment:
Are both the checks needed?

Reply:
the result might be overflow?

Reply:
@kripken the code just check overflow, dose it right?

Reply:
The code just needs to check overflow, correct. But I think one of the two is enough. It looks a little weird that it's not symmetric, that's true, but it's still correct I believe. Perhaps I'm missing something, though - can you find some values where both checks are needed to find the overflow?

Reply:
thanks, after the `#5038` is accepted. I will write more tests for this case, because now the parser of binaryen can not accept the number that is larger than int32_max.

Yifeng

Reply:
> The code just needs to check overflow, correct. But I think one of the two is enough. It looks a little weird that it's not symmetric, that's true, but it's still correct I believe. Perhaps I'm missing something, though - can you find some values where both checks are needed to find the overflow?

thanks, after the #5038 is accepted. I will write more tests for this case, because now the parser of binaryen can not accept the number that is larger than int32_max.
Yifeng

Reply:
Sorry, maybe I wasn't clear. This line can be simplified like this:
```cpp
bool overflow = base > max - offset;
```
It may be surprising that only one of the two checks you wrote is needed, but mathematically it works out. For simplicity, consider the 8-bit case, and that
```
base = 100
offset = 200
max = 255
```
Then
```
base > max - offset
100 > 255 - 200
100 > 55
true
```
and
```
offset > max - base
200 > 255 - 100
200 > 155
true
```
So the checks are redundant. We only need one of them.

Mathematically, `base > max - offset` and `offset > max - base` are equivalent. That is true both in normal math, where there are no overflows (just add offset and subtract base from the first equation to get the second), and also in wrapping math (both equations imply an overflow in `base + offset`, and an overflow there implies each of them, so we have equivalence).

Reply:
thanks, I add new test for the overflow check.